src/optimisation/Optimisation.jl

```diff
@@ -139,186 +154,364 @@ function (f::OptimLogDensity)(F, G, z)
+        OptimizationOptimJL.LBFGS()
     end
+end
 
-    d = length(vi[:])
-    lb = transform(fill(-Inf, d), vi, model, constrained_space{true}())
-    ub = transform(fill(Inf, d), vi, model, constrained_space{true}())
+"""
+    OptimizationProblem(log_density::OptimLogDensity, adtype, constraints)
 
-    return lb, ub
+Create an `OptimizationProblem` for the objective function defined by `log_density`.
+"""
+function Optimization.OptimizationProblem(log_density::OptimLogDensity, adtype, constraints)
+    # Note that OptimLogDensity is a callable that evaluates the model with given
+    # parameters. Hence we can use it in the objective function as below.
+    f = Optimization.OptimizationFunction(
+        (x, _log_density) -> _log_density(x), adtype; cons=constraints.cons
+    )
+    initial_params = log_density.varinfo[:]
+    prob = if !has_constraints(constraints)
+        Optimization.OptimizationProblem(f, initial_params, log_density)
+    else
+        Optimization.OptimizationProblem(
+            f, initial_params, log_density;
+            lcons=constraints.lcons,
+            ucons=constraints.ucons,
+            lb=constraints.lb,
+            ub=constraints.ub
+        )
+    end
+    return prob
 end
 
-function _optim_objective(model::DynamicPPL.Model, ::MAP, ::constrained_space{false})
-    ctx = OptimizationContext(DynamicPPL.DefaultContext())
-    obj = OptimLogDensity(model, ctx)
+"""
+    estimate_mode(
+        model::DynamicPPL.Model,
+        estimator::ModeEstimator,
+        [initial_params::Union{AbstractVector,Nothing}],
+        [solver];
+        <keyword arguments>
+    )
+
+Find the mode of the probability distribution of a model.
+
+Under the hood this function calls `Optimization.solve`.
+
+# Arguments
+- `model::DynamicPPL.Model`: The model for which to estimate the mode.
+- `estimator::ModeEstimator`: Can be either `MLE()` for maximum likelihood estimation or
+    `MAP()` for maximum a posteriori estimation.
+- `initial_params::Union{AbstractVector,Nothing}`: Initial value for the optimization.
+    Optional, unless non-box constraints are specified. If `nothing` or omitted it is
+    generated by either sampling from the prior distribution or uniformly from the box
+    constraints, if any.
+- `solver`. The optimization algorithm to use. Optional. Can be any solver recognised by
+    Optimization.jl. If `nothing` or omitted a default solver is used: LBFGS, or IPNewton if
+    non-box constraints are present.
+
+# Keyword arguments
+- `adtype::AbstractADType=AutoForwardDiff()`: The automatic differentiation type to use.
+- Keyword arguments `lb`, `ub`, `cons`, `lcons`, and `ucons` define constraints for the
+    optimization problem. Please see [`ModeEstimationConstraints`](@ref) for more details.
+- Any extra keyword arguments are passed to `Optimization.solve`.
+"""
+function estimate_mode(
+    model::DynamicPPL.Model,
+    estimator::ModeEstimator,
+    initial_params::Union{AbstractVector,Nothing},
```

Comment:
I'd suggest we make this a kwarg, similar to what it is for `sample`.

Reply:
Done.